Sources/KSCrashRecording/Monitors/KSCrashMonitor_Memory.m

```diff
@@ -0,0 +1,546 @@
+}
+
+/**
+ Mapping memory to a file on disk. This allows us to simply treat the location
+ in memory as a structure and the kernel will ensure it is on disk. This is also
+ crash resistant.
+ */
+static void ksmemory_map(const char* path)
+{
+    void *ptr = ksfu_mmap(path, sizeof(KSCrash_Memory));
+    if (!ptr) {
+        return;
+    }
+    
+    g_memory = (KSCrash_Memory *)ptr;
+    KSCrashAppMemory *memory = g_memoryTracker.memory;
+    
+    _ks_memory_update(^(KSCrash_Memory *mem) {
+        *mem = (KSCrash_Memory){
+            .footprint = memory.footprint,
+            .remaining = memory.remaining,
+            .pressure = (uint8_t)memory.pressure,
+            .level = (uint8_t)memory.level,
+            .limit = memory.limit,
+            .timestamp = ksdate_microseconds(),
+            .state = KSCrashAppStateTracker.shared.transitionState,
+            .fatal = false,
+        };
+    });
+}
+
+/**
+ What we're doing here is writing a file out that can be reused
+ on restart if the data shows us there was a memory issue.
+ 
+ If an OOM did happen, we'll modify this file
+ (see `kscm_memory_check_for_oom_in_previous_session`),
+ then write it back out using the normal writing procedure to write reports. This
+ leads to the system seeing the report as if it had always been there and will
+ then report an OOM.
+ */
+static void ksmemory_write_possible_oom(void)
+{
+    NSURL *reportURL = kscm_memory_oom_breadcrumb_URL();
+    const char *reportPath = reportURL.path.UTF8String;
+
+    KSMC_NEW_CONTEXT(machineContext);
+    ksmc_getContextForThread(ksthread_self(), machineContext, false);
+    KSStackCursor stackCursor;
+    kssc_initWithMachineContext(&stackCursor, KSSC_MAX_STACK_DEPTH, machineContext);
+    
+    char eventID[37] = {0};
+    ksid_generate(eventID);
+    
+    KSCrash_MonitorContext context;
+    memset(&context, 0, sizeof(context));
+    context.crashType = KSCrashMonitorTypeMemoryTermination;
+    context.eventID = eventID;
+    context.registersAreValid = false;
+    context.offendingMachineContext = machineContext;
+    context.currentSnapshotUserReported = true;
+    
+    // we don't need all the images, we have no stack
+    context.omitBinaryImages = true;
+    
+    // _reportPath_ only valid within this scope
+    context.reportPath = reportPath;
+
+    kscm_handleException(&context);
+}
+
+void ksmemory_initialize(const char* installPath)
+{
+    g_hasPostEnable = 0;
+    g_installURL = [NSURL fileURLWithPath:@(installPath)];
+    NSURL *dataURL = [g_installURL URLByAppendingPathComponent:@"Data"];
+    NSURL *memoryURL = [dataURL URLByAppendingPathComponent:@"memory"];
+    const char *path = memoryURL.path.UTF8String;
+    
+    // Make sure the folder exists
+    [[NSFileManager defaultManager] createDirectoryAtURL:dataURL withIntermediateDirectories:YES attributes:nil error:nil];
+    
+    // load up the old data
+    ksmemory_read(path);
+    
+    // map new data
+    ksmemory_map(path);
```

Comment:
Should we do mmap only on install?

Reply:
Memory is really a full process thing, you need the past to understand the present, so setting everything up very early is the way to go. 

`read` and `map` should stay together since read brings in the previous version of the data and map sets up the new one.

What do you think?

Reply:
You make a good point. Let's move it to on install.